llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-reka/llama_index/multi_modal_llms/reka/base.py

```diff
@@ -0,0 +1,447 @@
+from llama_index.core.schema import ImageDocument
+
+try:
+    from reka.client import Reka, AsyncReka
+    from reka.core import ApiError
+except ImportError:
+    raise ValueError(
+        "Reka is not installed. Please install it with `pip install reka-api`."
+    )
+
+DEFAULT_REKA_MODEL = "reka-flash"
+DEFAULT_REKA_MAX_TOKENS = 512
+DEFAULT_REKA_CONTEXT_WINDOW = 128000
+
+
+def process_messages_for_reka(messages: Sequence[ChatMessage]) -> List[Dict[str, str]]:
+    reka_messages = []
+    system_message = None
+
+    for message in messages:
+        if message.role == MessageRole.SYSTEM:
+            if system_message is None:
+                system_message = message.content
+            else:
+                raise ValueError("Multiple system messages are not supported.")
+        elif message.role == MessageRole.USER:
+            content = message.content
+            if system_message:
+                content = f"{system_message}\n{content}"
+                system_message = None
+            reka_messages.append({"role": "user", "content": content})
+        elif message.role == MessageRole.ASSISTANT:
+            reka_messages.append({"role": "assistant", "content": message.content})
+        else:
+            raise ValueError(f"Unsupported message role: {message.role}")
+
+    return reka_messages
+
+
+class RekaMultiModalLLM(MultiModalLLM):
+    """Reka Multi-Modal LLM integration for LlamaIndex."""
+
+    model: str = Field(default=DEFAULT_REKA_MODEL, description="The Reka model to use.")
+    temperature: float = Field(
+        default=DEFAULT_TEMPERATURE,
+        description="The temperature to use for sampling.",
+        gte=0.0,
+        lte=1.0,
+    )
+    max_tokens: int = Field(
+        default=DEFAULT_REKA_MAX_TOKENS,
+        description="The maximum number of tokens to generate.",
+        gt=0,
+    )
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict,
+        description="Additional keyword arguments for Reka API calls.",
+    )
+
+    _client: Reka = PrivateAttr()
+    _aclient: AsyncReka = PrivateAttr()
+
+    def __init__(
+        self,
+        model: str = DEFAULT_REKA_MODEL,
+        api_key: Optional[str] = None,
+        temperature: float = DEFAULT_TEMPERATURE,
+        max_tokens: int = DEFAULT_REKA_MAX_TOKENS,
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        additional_kwargs = additional_kwargs or {}
+        callback_manager = callback_manager or CallbackManager([])
+
+        api_key = api_key or os.getenv("REKA_API_KEY")
+        if not api_key:
+            raise ValueError(
+                "Reka API key is required. Please provide it as an argument or set the REKA_API_KEY environment variable."
+            )
+
+        self._client = Reka(api_key=api_key)
+        self._aclient = AsyncReka(api_key=api_key)
```

Comment:
as with the normal LLM, lets move this to after the init to avoid errors

Reply:
addressed.